src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -339,60 +345,93 @@ ISymbol memberSymbol
 			);
 		}
 
+		private enum AssignmentQueryKind {
+			NothingToCheck,
+			ImmutabilityQuery,
+			Hopeless
+		}
+
 		/// <summary>
 		/// For a field/property assignment, figure out what needs to be checked for it.
 		/// </summary>
 		/// <param name="assignment">The assignment syntax for the field/property (possibly null)</param>
-		/// <returns>null if no checks are needed, otherwise the query we need to run.</returns>
-		private ImmutabilityQuery? GetImmutabilityQueryForAssignment(
-			ExpressionSyntax assignment
+		/// <param name="query">The query to do (if the return value is AssignmentQueryKind.ImmutabilityQuery.)</param>
+		/// <param name="diagnostic">A diagnostic to report (if the return value is AssignmentQueryKind.Hopeless.)</param>
+		/// <returns>The details about what to check for this assignment</returns>
+		private AssignmentQueryKind GetQueryForAssignment(
+			ExpressionSyntax assignment,
+			out ImmutabilityQuery query,
+			out Diagnostic diagnostic
 		) {
 			// When we have an assignment we use it to narrow our check, e.g.
 			//
 			//   private readonly object m_lock = new object();
 			//
 			// is safe even though object (the type of the field) in general is not.
 
+			query = default;
+			diagnostic = null;
+
 			// Easy cases
 			switch( assignment.Kind() ) {
-				// Sometimes people explicitly (but needlessly) assign null in
-				// an initializer... this is safe.
 				case SyntaxKind.NullLiteralExpression:
+					// Sometimes people explicitly (but needlessly) assign null in
+					// an initializer... this is safe.
+					return AssignmentQueryKind.NothingToCheck;
 
-				// Lambda initializers for readonly members are safe
-				// because they can only close over other members, which
-				// will be checked independently, or static members of
-				// another class, which are also analyzed
-				//
-				// TODO: only true (in general) for initializers; non-static
-				// ones in constructors can close over mutable stack vars!
 				case SyntaxKind.SimpleLambdaExpression:
 				case SyntaxKind.ParenthesizedLambdaExpression:
-					return null;
-			}
-
-			var model = m_compilation.GetSemanticModel( assignment.SyntaxTree );
-			var typeInfo = model.GetTypeInfo( assignment );
+					if ( assignment.Parent.Parent is VariableDeclaratorSyntax ) {
+						// Lambda initializers for readonly members are safe
+						// because they can only close over other members,
+						// which will be checked independently, or static
+						// members of another class, which are also analyzed.
+						return AssignmentQueryKind.NothingToCheck;
+					}

```

Comment:
I thiiiink this might also match?

```csharp
class C {
    private readonly Func<int> m_foo;
    
    public C() {
        Func<int> bar = m_foo = () => 0;
    }
    

```